src/Accounts/Authentication/KeyStore/AzKeyStore.cs

```diff
@@ -11,67 +11,220 @@
-using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
-using Microsoft.WindowsAzure.Commands.Common;
+using Newtonsoft.Json;
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
-using System.Security;
+using System.Text;
 
 namespace Microsoft.Azure.Commands.ResourceManager.Common
 {
     public class AzKeyStore : IDisposable
     {
         public const string Name = "AzKeyStore";
-        private static readonly IDictionary<IKeyStoreKey, SecureString> _credentials = new Dictionary<IKeyStoreKey, SecureString>();
 
-        [Obsolete("The constructor is deprecated. Will read key from encryted storage later.", false)]
-        public AzKeyStore(IAzureContextContainer profile)
+        internal class KeyStoreElement
+        {
+            public string keyType;
+            public string keyStoreKey;
+            public string valueType;
+            public string keyStoreValue;
+        }
+
+        private static IDictionary<Type, string> _typeNameMap = new ConcurrentDictionary<Type, string>();
+
+        private static IDictionary<string, JsonConverter> _elementConverterMap = new ConcurrentDictionary<string, JsonConverter>();
+
+        public static void RegisterJsonConverter(Type type, string typeName, JsonConverter converter = null)
+        {
+            if (string.IsNullOrEmpty(typeName))
+            {
+                throw new ArgumentNullException($"typeName cannot be empty.");
+            }
+            if (_typeNameMap.ContainsKey(type))
+            {
+                if (string.Compare(_typeNameMap[type], typeName) != 0)
+                {
+                    throw new ArgumentException($"{typeName} has conflict with {_typeNameMap[type]} with reference to {type}.");
+                }
+            }
+            else
+            {
+                _typeNameMap[type] = typeName;
+            }
+            if (converter != null)
+            {
+                _elementConverterMap[_typeNameMap[type]] = converter;
+            }
+        }
+
+        private IDictionary<IKeyStoreKey, Object> _credentials = new ConcurrentDictionary<IKeyStoreKey, Object>();
+        private IStorage _storage = null;
+
+        private bool autoSave = true;
+        private Exception lastError = null;
+
+        public IStorage Storage
+        {
+            get => _storage;
+            set => _storage = value;
+        }
+
+        public AzKeyStore()
+        {
+
+        }
+
+        public AzKeyStore(string directory, string fileName, bool loadStorage = true, bool autoSaveEnabled = true, IStorage inputStorage = null)
+        {
+            autoSave = autoSaveEnabled;
+            Storage = inputStorage ?? new StorageWrapper()
+            {
+                FileName = fileName,
+                Directory = directory
+            };
+            Storage.Create();
+
+            if (loadStorage)
+            {
+                if (!LoadStorage())
```

Comment:
is this an async operation?

Reply:
It is a lightweight storage. I think we can sync mode here.